lib/nodes/opendss.cpp

```diff
@@ -0,0 +1,378 @@
+      throw SystemError("Invalid monitor name '{}'", m_name);
+    }
+  }
+
+  return Node::prepare();
+}
+
+int OpenDSS::start() {
+
+  //Start with writing
+  writing_turn = true;
+
+  return Node::start();
+}
+
+int OpenDSS::extractMonitorData(struct Sample *const *smps) {
+
+  // Get the data from the OpenDSS monitor
+  uintptr_t myPtr;
+  int myType;
+  int mySize;
+  int data_count = 0;
+  for (auto& Name:monitor_name) {
+    MonitorsS(2, Name.data());
+    MonitorsV(1, &myPtr, &myType, &mySize);
+
+    int channel = MonitorsI(17, 0);
+    if (data_count == 0) {
+      channel += 2;
+    }
+
+    float* data_ptr = reinterpret_cast<float*>(myPtr);
+    data_ptr += (mySize / 4) - channel;
+    for (int i = 0; i < channel; i++) {
+      smps[0]->data[data_count + i].f = *data_ptr;
+      data_ptr++;
+    }
+    data_count += channel;
+  }
+  return data_count;
+}
+
+int OpenDSS::_read(struct Sample *smps[], unsigned cnt) {
+
+  int ret = 0;
+  // Wait until writing is done
+  pthread_mutex_lock(&mutex);
+  while (writing_turn) {
+    pthread_cond_wait(&cv, &mutex);
+  }
+
+  smps[0]->ts.origin = ts;
+  smps[0]->flags = (int)SampleFlags::HAS_DATA | (int)SampleFlags::HAS_TS_ORIGIN | (int)SampleFlags::HAS_SEQUENCE;
+  // smps[0]->signals = n->getInputSignals(false);
+  smps[0]->length = 0;
+
+  // Solve OpenDSS file
+  SolutionI(0, 0);
+
+  smps[0]->length = extractMonitorData(smps);
+  smps[0]->sequence = smps[0]->data[0].f;
+
+  ret = 1;
+  writing_turn = true;
+  pthread_cond_signal(&cv);
+  pthread_mutex_unlock(&mutex);
+
+  return ret;
+}
+
+int OpenDSS::_write(struct Sample *smps[], unsigned cnt) {
+
+  // Wait until reading is done
+  pthread_mutex_lock(&mutex);
+  while (!writing_turn) {
+    pthread_cond_wait(&cv, &mutex);
+  }
+
+  ts = smps[0]->ts.origin;
+  int i = 0;
+
+  for (auto &ele:dataIn) {
+    double (*func)(int, double);
+    switch (ele.type) {
+      case ElementType::generator:
+        func = GeneratorsF;
+        GeneratorsS(1, ele.name.data());
+        break;
+      case ElementType::load:
+        func = DSSLoadsF;
+        DSSLoadsS(1, ele.name.data());
+        break;
+      case ElementType::isource:
+        func = IsourceF;
+        IsourceS(1, ele.name.data());
+        break;
+      default:
+        throw SystemError("Invalid element type");
+    }
+
+    for (auto &m: ele.mode) {
+      func(m, smps[0]->data[i].f);
+      i++;
+    }
+  }
+
+  writing_turn = false;
+  pthread_cond_signal(&cv);
+  pthread_mutex_unlock(&mutex);
+
+  return cnt;
+}
+
+int OpenDSS::stop() {
+
+  // Close OpenDSS
+  cmd_command = "CloseDI";
+	cmd_result = DSSPut_Command(cmd_command.data());
+
+  return Node::stop();
+}
+
+// Register node
+static char n[] = "opendss";
+static char d[] = "A node providing interface with OpenDSS";
+static NodePlugin<OpenDSS, n, d,
+                  (int)NodeFactory::Flags::SUPPORTS_READ |
+                      (int)NodeFactory::Flags::SUPPORTS_WRITE,
+                      1>
+    p;
```

Comment:
Did you format this PR already with clang-format?

```suggestion
static NodePlugin<OpenDSS, n, d,
                  (int)NodeFactory::Flags::SUPPORTS_READ |
                  (int)NodeFactory::Flags::SUPPORTS_WRITE, 1>
    p;
```